File: baekjoon/java/src/silver2/Q24913_Vote.java
package silver2;
// https://www.acmicpc.net/problem/24913

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Q24913_Vote {
    static FastReader scan = new FastReader();
    static StringBuffer sb = new StringBuffer();

    static void input() {
        N = scan.nextInt();
        Q = scan.nextInt();
        res = new long[N + 1];
    }

    static int N, Q;
    static long jh = 0;
    static long maxV = 0;
    static long sum = 0;
    static long[] res;

    static void sol() {
        for (int i = 0; i < Q; i++) {
            int op = scan.nextInt();
            if (op == 1) {
                int cnt = scan.nextInt();
                int cand = scan.nextInt();
                if (cand != N + 1) {
                    res[cand] += cnt;
                    maxV = Math.max(maxV, res[cand]);
                    sum += cnt;
                } else {
                    jh += cnt;
                }
            } else if (op == 2) {
                int me = scan.nextInt();
                int other = scan.nextInt();

                if (jh + me <= maxV) {
                    sb.append("NO").append("\n");
                } else {
                    long toCompare = Math.max(maxV, (long)Math.ceil((sum + other) / (double)N));
                    if(jh+me > toCompare) {
                        sb.append("YES").append("\n");
                    } else {
                        sb.append("NO").append("\n");
                    }
                }
            }
        }

    }

    public static void main(String[] args) {
        input();
        sol();
        System.out.println(sb.toString());
    }

    static class FastReader {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        String next() {
            while (st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }
    }
}

File: baekjoon/python/silver2/q24913_vote.py
import sys
input = sys.stdin.readline

n, q = map(int, input().split())

result = [0] * (n+2)
print(result)

max_v = 0
sum = 0

for _ in range(q):
    q, x, y =  map(int, input().split())
    if q == 1:
        result[y] += x
        if y != n+1:
            max_v = max(max_v, result[y])
            sum += x
    
    else:
        print(result, max_v, sum)
        if result[n+1] + x <= max(result[1:n+1]):
            print("NO")
        else:
            if 
            # 더해야 할 것이 y 
            print("@@@")
            to_add = max_v * n - sum

            if to_add <= y:
                print("YES")
            else:
                to_add -= y
                res = to_add // n
                if to_add % n > 0:
                    res += 1
                if max_v + res < result[n+1]+x:
                    print("YES")
                else:
                    print("NO")

